gluten-uniffle/src/main/java/org/apache/spark/shuffle/gluten/uniffle/UniffleShuffleManager.java

```diff
@@ -0,0 +1,193 @@
+      String name, SparkConf conf, boolean isDriver) {
+    Constructor constructor;
+    ColumnarShuffleManager instance;
+    try {
+      Class klass = Class.forName(name);
+      try {
+        constructor = klass.getConstructor(conf.getClass(), Boolean.TYPE);
+        instance = (ColumnarShuffleManager)constructor.newInstance(conf, isDriver);
+      } catch (NoSuchMethodException var7) {
+        constructor = klass.getConstructor(conf.getClass());
+        instance = (ColumnarShuffleManager)constructor.newInstance(conf);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("initColumnManager fail");
+    }
+    return instance;
+  }
+
+  private boolean isDriver() {
+    return "driver".equals(SparkEnv.get().executorId());
+  }
+
+  // This method is called in Spark driver side,
+  // and Spark driver will make some decision according to coordinator,
+  // e.g. determining what RSS servers to use.
+  // Then Spark driver will return a ShuffleHandle and
+  // pass that ShuffleHandle to executors (getWriter/getReader)
+  @Override
+  public <K, V, C> ShuffleHandle registerShuffle(
+      int shuffleId, ShuffleDependency<K, V, C> dependency) {
+    if (!(dependency instanceof ColumnarShuffleDependency)) {
+      columnarShuffleIds.add(shuffleId);
+      return columnarShuffleManager().registerShuffle(shuffleId, dependency);
+    }
+    return super.registerShuffle(shuffleId, dependency);
+  }
+
+  @Override
+  public <K, V> ShuffleWriter<K, V> getWriter(
+      ShuffleHandle handle,
+      long mapId,
+      TaskContext context,
+      ShuffleWriteMetricsReporter metrics) {
+    if (!(handle instanceof RssShuffleHandle)) {
+      columnarShuffleIds.add(handle.shuffleId());
+      return columnarShuffleManager().getWriter(handle, mapId, context, metrics);
+    }
+    RssShuffleHandle<K, V, V> rssHandle = ((RssShuffleHandle<K, V, V>) handle);
+    // todo: this implement is tricky, we should refactor it
+    if (id.get() == null) {
+      id.compareAndSet(null, rssHandle.getAppId());
+      dataPusher.setRssAppId(id.get());
+    }
+    int shuffleId = rssHandle.getShuffleId();
+    String taskId = "" + context.taskAttemptId() + "_" + context.attemptNumber();
+    BufferManagerOptions bufferOptions = new BufferManagerOptions(sparkConf);
+    ShuffleWriteMetrics writeMetrics;
+    if (metrics != null) {
+      writeMetrics = new RssShuffleManager.WriteMetrics(metrics);
+    } else {
+      writeMetrics = context.taskMetrics().shuffleWriteMetrics();
+    }
+    WriteBufferManager bufferManager = new WriteBufferManager(
+        shuffleId, taskId, context.taskAttemptId(), bufferOptions,
+        null,
+        rssHandle.getPartitionToServers(), context.taskMemoryManager(),
+        writeMetrics, RssSparkConfig.toRssConf(sparkConf), this::sendData);
+    LOG.info("RssHandle appId {} shuffleId {} ", rssHandle.getAppId(), rssHandle.getShuffleId());
+    return new UniffleColumnarShuffleWriter<>(
```

Comment:
why not  UniffleColumnarShuffleWriter(super.getWriter(),...) .   Too much RSS code is exposed in gluten

Reply:
Sounds reasonable, but  `rss.row.based` have to be set in `WriteBufferManager`, Though it looks really not elegant